### PR TITLE
8282209: G1: Use synchronous concurrent marking for the code cache

### PR DESCRIPTION
**Pull request by lkorinth**
This fix removes collect_for_codecache(), and uses collect() directly instead. This has the advantage that the g1 code path can be run concurrently using a concurrent mark instead of a full gc.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - Committer)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/loom pull/102/head:pull/102` \
`$ git checkout pull/102`

Update a local copy of the PR: \
`$ git checkout pull/102` \
`$ git pull https://git.openjdk.java.net/loom pull/102/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 102`

View PR using the GUI difftool: \
`$ git pr show -t 102`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/loom/pull/102.diff">https://git.openjdk.java.net/loom/pull/102.diff</a>

</details>

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good. Ship it!

File: src/hotspot/share/gc/g1/g1FullCollector.cpp [hunk 1]
```diff
@@ -188,15 +188,15 @@ void G1FullCollector::prepare_collection() {
...
-    // safepoint.
-    CodeCache::finish_marking_cycle();
+  if (!CodeCache::is_marking_cycle_active()) {
+    // This is the normal case when we do not call collect when a
+    // concurrent mark is ongoing. We then start a new code marking
+    // cycle. If, on the other hand, a concurrent mark is ongoing, we
+    // will be conservative and use the last code marking cycle. Code
+    // caches marked between the two concurrent marks will live a bit
+    // longer than needed.
+    CodeCache::start_marking_cycle();
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1FullCollector.cpp:198]: Please add a helper method for this copy&paste in `G1CollectedHeap`. Something like `start_codecache_marking_cycle_if_inactive` (not perfect) or similar.

File: src/hotspot/share/gc/g1/g1Policy.cpp [hunk 2]
```diff
@@ -1080,6 +1080,7 @@ void G1Policy::decide_on_concurrent_start_pause() {
       initiate_conc_mark();
       log_debug(gc, ergo)("Initiate concurrent cycle (concurrent cycle initiation requested)");
     } else if (_g1h->is_user_requested_concurrent_full_gc(cause) ||
+               (cause == GCCause::_codecache_GC_threshold) ||
                (cause == GCCause::_wb_breakpoint)) {
       // Initiate a user requested concurrent start or run to a breakpoint.
       // A concurrent start must be young only GC, so the collector state
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1Policy.cpp:1086]: The comment needs an update.
to

**Review by tschatzl — CHANGES_REQUESTED** [review]

**Review by tschatzl — APPROVED** [review]
Lgtm. Thanks.

**Review by fisk — APPROVED** [review]
Looks good. Ship it!

**Pull request closed without merging**